feat: cqrs 기능 구현

kimtaewoo9 opened the pull request:
## 개요
성능을 위해 쓰기 기능과 조회 기능을 분리하였습니다.
로직의 분리 뿐만 아니라, 성능 개선을 위해 데이터 베이스도 분리하였습니다.

<img width="400" height="367" alt="image" src="https://github.com/user-attachments/assets/5bbdba8c-7d8f-49a3-aa91-14ac619dfa86" />


피드 조회 특성상, 최신 1000개 정도의 데이터에 접근하는 비율이 압도적으로 높을 것으로 예상하였습니다.
최신 1000개의 데이터를 역정규화하여 redis 에 캐시해놓고, 최신 데이터 1000개가 아닌 데이터에 대해서는 RDBMS에서 원본 데이터를 요청하여 Query Model 을 만들어 redis에 캐시해놓고 해당 데이터를 반환하도록 구현하였습니다.

데이터 동기화 문제는 카프카를 사용하여 실시간으로 Query 쪽 데이터베이스의 내용을 수정하도록 구현하였습니다.
<img width="865" height="318" alt="image" src="https://github.com/user-attachments/assets/69ed63af-29e0-4ed6-b213-6bf6f79442ba" />

## 피드 목록 조회 최적화
<img width="877" height="298" alt="image" src="https://github.com/user-attachments/assets/923afad0-9f5e-43f4-afe5-1222424ef134" />

피드 목록 조회 서비스는 Kafka로 부터 전달 받은 피드 생성/삭제 이벤트로 Redis에 피드 목록을 저장하였습니다. Client가 피드 목록을 요청한다면 최신 1000개 이내의 목록 데이터는 Redis 에서 가져오고, 이후의 데이터는 피드 서비스에서 가져오도록 구현하였습니다.

항상 최신순 정렬을 유지하기 위해, Score 값을 해당 Feed의 생성 시간(createdAt)으로 넣어주었습니다. (Redis의 **SortedSet** 기능을 활용하여 최신순 정렬)

## 특이사항

QuokkaJoa committed:
feat: 의상 추천을 위한 메서드 구현 구체화

QuokkaJoa committed:
merge 커밋

Commit:
fix: 데이터 조회 기능 위도, 경도에서 좌표로 변경

Commit:
fix: 현재 위치로 설정하기 버튼시 날짜 최신 버전 안가져 오는 문제 수정

QuokkaJoa committed:
feat: sql 파라미터 trace 추가

QuokkaJoa committed:
fix: 오타 수정

QuokkaJoa committed:
fix: 불러오는 데이터 수정(전일비교온도 -> 현재 온도)

QuokkaJoa committed:
feat: 점수를 기반으로 한 추천 기능 추가(기틀)

QuokkaJoa committed:
merge 커밋

Commit:
feat: 날씨 변경시 알림기능 추가

Commit:
feat: 중복 알림 전송하지 못하는 기능 추가

kimtaewoo9 committed:
feat: 충돌 해결

kimtaewoo9 committed:
feat: 충돌 해결

QuokkaJoa committed:
feat: 점수별로 최적의 옷을 추천해주는 옷 추천 기능 자체 알고리즘 구현

QuokkaJoa committed:
merge 커밋

kimtaewoo9 committed:
feat: cqrs 기능 구현 (각 이벤트의 payload 만들기, query model repository 만들기, 조회 서비스 만들기, command service, query service 나누기)

kimtaewoo9 committed:
feat: cqrs 기능 구현 (각 이벤트의 payload 만들기, query model repository 만들기, 조회 서비스 만들기, command service, query service 나누기)

kimtaewoo9 committed:
feat: 주석 수정

kimtaewoo9 committed:
feat: 주석 수정

kimtaewoo9 committed:
feat: test 코드 작성

kimtaewoo9 committed:
feat: test 코드 작성

kimtaewoo9 committed:
refactor: 오류 수정

kimtaewoo9 committed:
refactor: 오류 수정

kimtaewoo9 committed:
refactor: local kafka 로 연결하도록 수정

kimtaewoo9 committed:
refactor: local kafka 로 연결하도록 수정

kimtaewoo9 committed:
refactor: local kafka 로 연결하도록 수정 (#174)

kimtaewoo9 committed:
refactor: readAll 메서드 오류 수정

kimtaewoo9 committed:
refactor: readAll 메서드 오류 수정

kimtaewoo9 committed:
refactor: readAll 메서드 오류 수정

kimtaewoo9 committed:
refactor: readAll 메서드 오류 수정

kimtaewoo9 committed:
refactor: Message Relay 위치 수정

kimtaewoo9 committed:
refactor: Message Relay 위치 수정

urur-27 reviewed (approved):
고생하셨습니다! 변경점이 많아 모두 확인하진 못했으나 작성해주신 내용을 바탕으로 redis, kafka를 이용해서 처리한다는 점 확인했습니다.

urur-27 committed:
feat(clothes): WebFlux 전환 및 JPA LAZY 예외 해소를 위한 스냅샷 패턴 도입

- ClothingExtractionService: blocking 구간을 boundedElastic로 분리
  - HtmlExtractor.extract() (Jsoup) → fromCallable + subscribeOn
  - AttributeReadService.loadAllForExtraction() (JPA) → fromCallable + subscribeOn
- AttributeReadService: @Transactional(readOnly=true) + EntityGraph로 옵션까지 로드
  - 엔티티 → record 스냅샷(AttributeDefSnap/AttributeSnapshot) 변환
  - UUID 기반 ID 정합성 일치
- AttributeMapper: Repository 의존 제거, 스냅샷 기반 순수 매핑
- 중복 DB 조회 제거, WebFlux 체인에서 DTO만 전달 → LazyInitializationException 해소
- LLM 요청: idempotency-key 헤더, timeout(35s), retry(backoff 250ms, 1회) 적용
- 테스트 코드 임시 비활성화

urur-27 committed:
chore: 사용하지 않는 파일 삭제

urur-27 committed:
pull 충돌 해결

Commit:
Feat: OAuth 기능 구현 및 Authentication Manager, Provider 구현 수정
Fix: api /me 수정 구현 -> 새로고침 시 원래 페이지로 돌아옴

Commit:
Merge branch 'develop' into feat/user-oauth

QuokkaJoa committed:
merge 커밋

urur-27 committed:
fix: 날씨 정보 조회 버그 수정 (#168)

urur-27 committed:
feat: LLM 프록시 서버 추가 및 Spring WebFlux 기반 비동기 처리 적용 (#175)

urur-27 committed:
Feat: OAuth 기능 구현 (#176)

QuokkaJoa committed:
merge 커밋

QuokkaJoa committed:
feat: 날씨 추천 알고리즘 변경(점수제) (#177)

kimtaewoo9 committed:
refactor: DBMS, Redis 동기화 오류 수정

kimtaewoo9 committed:
refactor: 오류 수정(CustomUserDetails 패키지 위치)

kimtaewoo9 committed:
refactor: 충돌 해결

kimtaewoo9 committed:
refactor: 충돌 해결

kimtaewoo9 committed:
refactor: test 코드 오류 수정

kimtaewoo9 committed:
refactor: test 코드 패키지 위치 수정
